File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        res = []
        def dfs_pre_order(root):
            if not root:
                res.append('N')
                return 
            res.append(str(root.val))
            dfs_pre_order(root.left)
            dfs_pre_order(root.right)
        dfs_pre_order(root)
        return ','.join(res)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        
        def dfs_pre_order():
            nonlocal i
            i += 1
            if data[i] == 'N':
                return
            root = TreeNode(data[i])
            root.left = dfs_pre_order()
            root.right = dfs_pre_order()
            return root
            
        i = -1
        print(data)
        data = data.split(',')
        return dfs_pre_order()
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {
    StringBuilder res;
    int i;
    public String dfs(TreeNode root, String res) {
        if (root == null) {
            res += "N,";
        } else {
        res += res.valueOf(root.val) + ",";
        res = dfs(root.left, res);
        res = dfs(root.right, res);
        }
        return res;
    }

    public TreeNode buildTree(List<String> l) {
    // Recursive deserialization.
    if (l.get(0).equals("N")) {
      l.remove(0);
      return null;
    }

    TreeNode root = new TreeNode(Integer.valueOf(l.get(0)));
    l.remove(0);
    root.left = buildTree(l);
    root.right = buildTree(l);

    return root;
        
    }
    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        return dfs(root, "");
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] values = data.split(",");
        List<String> data_list = new LinkedList<String>(Arrays.asList(values));
        i = -1;
        return buildTree(data_list);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));